lib/sqlalchemy/orm/mapped_collection.py

```diff
@@ -315,21 +353,30 @@ def __init__(self, keyfunc, *, ignore_unpopulated_attribute=False):
         self.ignore_unpopulated_attribute = ignore_unpopulated_attribute
 
     @classmethod
-    def _unreduce(cls, keyfunc, values):
-        mp = KeyFuncDict(keyfunc)
+    def _unreduce(
+        cls, keyfunc: Callable[[_KT], _KT], values: Dict[_KT, _KT]
+    ) -> "KeyFuncDict[_KT, _KT]":
+        mp: KeyFuncDict[_KT, _KT] = KeyFuncDict(keyfunc)
         mp.update(values)
         return mp
 
-    def __reduce__(self):
+    def __reduce__(
+        self,
+    ) -> tuple[
+        Callable[[_KT, _KT], KeyFuncDict[_KT, _KT]],
+        tuple[Any, dict[_KT, _KT] | dict[_KT, _KT]],
```

Comment:
```| dict[_KT, _KT]``` looks redundant.

Reply:
Also `|` is not supported by Python 3.7 which SQLA must support. So only `Union[]` can be used.